subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1268 @@
+					   sizeof(struct audio_module_message));
+		if (ret) {
+			data_fifo_block_free(rx_handle->thread.msg_rx, (void **)&data_msg_rx);
+
+			LOG_WRN("Module %s failed to queue audio data, ret %d", rx_handle->name,
+				ret);
+			return ret;
+		}
+
+		LOG_DBG("Audio data sent to module %s", rx_handle->name);
+
+	} else {
+		LOG_WRN("Receiving module %s is in an invalid state %d", rx_handle->name,
+			rx_handle->state);
+		return -ECANCELED;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Send audio data item to the module's TX FIFO.
+ *
+ * @param handle[in/out]  The handle for this modules instance.
+ * @param audio_data[in]  A pointer to the audio data.
+ *
+ * @return 0 if successful, error otherwise.
+ */
+static int tx_fifo_put(struct audio_module_handle *handle,
+		       struct audio_data const *const audio_data)
+{
+	int ret;
+	struct audio_module_message *data_msg_tx;
+
+	/* Take a TX message. */
+	ret = data_fifo_pointer_first_vacant_get(handle->thread.msg_tx, (void **)&data_msg_tx,
+						 K_NO_WAIT);
+	if (ret) {
+		LOG_WRN("No free space in TX FIFO for module %s, ret %d", handle->name, ret);
+		return ret;
+	}
+
+	/* Configure audio data. */
+	memcpy(&data_msg_tx->audio_data, audio_data, sizeof(struct audio_data));
+	data_msg_tx->tx_handle = handle;
+	data_msg_tx->response_cb = audio_data_release_cb;
+
+	/* Send audio data to modules output message queue. */
+	ret = data_fifo_block_lock(handle->thread.msg_tx, (void **)&data_msg_tx,
+				   sizeof(struct audio_module_message));
+	if (ret) {
+		LOG_ERR("Failed to send audio data to output of module %s, ret %d", handle->name,
+			ret);
+
+		data_fifo_block_free(handle->thread.msg_tx, (void **)&data_msg_tx);
+
+		ret = k_sem_take(&handle->sem, K_NO_WAIT);
+		if (ret) {
+			LOG_ERR("Failed to take semaphore for data release callback function");
+		}
+
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Send the audio data item to all connected modules.
+ *
+ * @param handle[in/out]  The handle for this modules instance.
+ * @param audio_data[in]  A pointer to the audio data.
+ *
+ * @return 0 if successful, error otherwise.
+ */
+static int send_to_connected_modules(struct audio_module_handle *handle,
+				     struct audio_data const *const audio_data)
+{
+	int ret;
+	struct audio_module_handle *handle_to;
+
+	if (handle->dest_count == 0) {
+		LOG_DBG("Nowhere to send the audio data from module %s so releasing it",
+			handle->name);
+
+		k_mem_slab_free(handle->thread.data_slab, (void **)audio_data->data);
+
+		return 0;
+	}
+
+	/* We need to ensure that all receiving modules have got the audio data.
+	 * This is so the first receiver cannot free the audio data before all receivers
+	 * have all gotten the audio data.
+	 */
+	ret = k_mutex_lock(&handle->dest_mutex, LOCK_TIMEOUT_US);
+	if (ret) {
+		LOG_ERR("Failed to take MUTEX lock in time");
+		return ret;
+	}
+
+	ret = k_sem_init(&handle->sem, handle->dest_count, handle->dest_count);
```

Comment:
To me it seems strange that a semaphore is inited like this in a function that is continously called

Reply:
Yes it is initiated here so that it gives a count of the number of times the data is sent (saves copying) and thus needs to align with the current number of connections.

Can we discuss?